### ISSUE TITLE
Investigate improving YSQL apps that have large and/or frequent number of connections

### ISSUE THREAD

**Issue by rkarthik007** [yugabyte/yugabyte-db#7985]
Jira Link: [DB-2375](https://yugabyte.atlassian.net/browse/DB-2375)
We can initially try how the following help (or not):
* pgbouncer
* pgpool

**Comment by fritshoogland-yugabyte** [comment]
Problem statement:

1. expensive to bootstrap a new process
1.1 Ysql/postgres is a database server that creates a dedicated process for each client connection, and removes the process when the client connection is terminated. 
1.2 A new database connection must be bootstrapped/initialized, which means it needs to lookup information.
1.3 Each database connection requires an amount of memory, even if it's never used.
1.4 A lot of connectionpools can be defined to start with a certain minimum number of connections, and grow when there's more activity, and decrease the number when activity becomes less, generating smaller or larger connection storms in the sense of adding connections or removing connections.
1.5 Some applications don't do connection management at all, and thus logon and use a new database connection, execute the database task and then remove the connection. Needless to say, this adds significant latency to database execution because of the initialization that is repeatedly needed on multiple levels.

2. too much connections
2.1 It's fairly normal for database applications to startup a large number of connections (including or excluding connection pools), probably to overcome the issue of running out of available connections and thus connection management, moving the problem towards the database. 
2.1.1 In a lot of cases the number of connections per application server is high enough to have a number of connections that is never used, and in a lot of cases there are multiple up to numerous application servers, each with a high number of connections.

3. resource management
3.1 Because each YSQL connection requires memory to be allocated unique to the connection, allocating a high number of connections will take a certain amount of memory, and given enough connections it can influence the availability of memory for master and tserver processing.
3.2 Obtaining connections quite often is batched. In certain cases this makes sense, such as application startup. However if this is done during normal running, batching new connections might influence running activity, which might cause these applications to spawn new connections, generating a connection storm.
3.3 When running with the high amounts of database connections, there is a chance the application server or application servers truly try to execute a database transaction in all the created connections. In a lot of cases, this will kill the server in the sense of resources, leading to immense queueing.

**Comment by fritshoogland-yugabyte** [comment]
pgpool (https://www.pgpool.net/mediawiki/index.php/Main_Page)
1. connection pooling
Each client connection to pgpool creates a connection from the pool to a defined postgres server. 
The connection between the pool and the postgres server can be maintained, so that a new client to pgpool can reuse it.
The client to database connection always is a 1:1 relationship.
New with version 4.1 is the option to do statement level load balancing.
Queries are parsed by pgpool to determine if it's a DML(+DDL?) "write" query or a "read" query. 
Read queries can be run on any database in the pool, write queries only on the designated primary database.
2. replication
pgpool can send changes to multiple (database) targets, providing replication.
3. load balancing
By default, the pool's mode is session, which means a session gets load balanced at connection time.
With the new statement level pool mode, a session can be load balanced per statement. Obviously for read statements only.
4. limit connection count
Postgres returns an error to a client that wants to connection when it's higher than the set maximum (postgresql.conf:max_connections). pgpool has a limit too, but it will queue new connections.
5. watchdog
pgpool high availability, in the case of multiple pgpool servers.
6. in memory cache
because pgpool parses each query before it's (potentially) handed to a postgres server, it understands what the query does. The first determination is whether a query reads or writes, but it could also keep a cache of data. For yugabyte where multiple tservers serve the same database, to me this looks dangerous for data consistency.

pgpool (pgpool-II) is a postgres connection pool intended or originally intended to handle connections towards a postgres database that has a standby database to provide high availability. This is most visible with the limitation that a write query always will be directed to the designated primary node. 

pgpool does provide server load balancing, which pgbouncer doesn't.

**Comment by fritshoogland-yugabyte** [comment]
pgbouncer (https://www.pgbouncer.org)

1. connection pooling
Each client connection to pgbouncer creates a connection to pgbouncer. The number of connections from pgbouncer to a defined postgres server can be defined independently and differently. This allows an "asymetrical setup" where the clients can use as many connections as they like, but the number of connections in YSQL is configured independently and potentially different.
pgbouncer keeps a list of database clients and database servers for which the relationship is defined in a pool, and connects the two based on the pool's properties.
2. no load balancing
pg bouncer does not do load balancing. clients connection to a pool in pgbouncer, which has a single database destination configured. load balancing can be done with another product, haproxy is often used (https://www.haproxy.com).
3. front end / back end pool
pg bouncer can be installed close to the application in order to lower the TCP roundtrip time for connections to be setup (front end), or close to the database when multiple clients inefficiently connect to the database so their inefficiency influences the database connections as little as possible.
4. most optimal usage
pgbouncer can most optimally use database connections when the pool property is set to transaction, so a postgres/YSQL connection is taken for the duration of a transaction. other pool options are 'statement' and 'session'. 
5. caveat: prepared statements
because of the simplicity of pgbouncer, there is no tracking of (database) server side properties such as prepared statements. This means the pool options 'transaction' and 'statement' can not be used when prepared statements are used. setting the pool mode to session means each connection will require its own database server connection.

**Comment by fritshoogland-yugabyte** [comment]
Each independent connection-pool product sitting in between a client and the database server will add latency, because it adds an extra network 'hop'. This means that an ideal situation would mean a database client handles its connections efficiently itself, not requiring an additional product. Efficient means:
(postgres and yugabyte)
- keep the number of database connections low.
- queue new connections if more active connections are needed.
- keep database connections connected, never build the number of connections up and down.
- reuse the same connection for identical queries (so the server side cache's can most optimally be used).

For yugabyte only:
- understand which other stateless postgres servers are available, and track their state (so availability and maybe load is known).
- fail over to another server if the current server becomes unavailable (probably via TCP timeout).

**Comment by ddorian** [comment]
Here is a post on analyzing connection scalability problems in PostgreSQL: https://www.citusdata.com/blog/2020/10/08/analyzing-connection-scalability/

A part of it has been committed to PG14: https://techcommunity.microsoft.com/t5/azure-database-for-postgresql/improving-postgres-connection-scalability-snapshots/ba-p/1806462 which we can try to commit too ?


> The most efficient use is transaction pooling,

I'm pretty sure it's statement pooling the most efficient, no ?

**Comment by fritshoogland-yugabyte** [comment]
Yes, statement pooling is most efficient, but this doesn’t work with multi statement transactions, which I think is not acceptable.

**Comment by fritshoogland-yugabyte** [comment]
+oddyssey

**Comment by ddorian** [comment]
There is 1 product that I know off that offers a connection pooler built in, PostgresPro Enterprise: https://postgrespro.com/docs/enterprise/12/connection-pooling ([limitations](https://postgrespro.com/docs/enterprise/12/connection-pooler-limitations), [how it works](https://postgrespro.com/docs/enterprise/12/how-connection-pooler-works)). Looks like they chose the transaction pooling method.

**Comment by ddorian** [comment]
In PostgreSQL 14, there's also committed [Batch / Pipeline Mode in libpq](https://www.postgresql.org/message-id/flat/CAJkzx4TXG18%2BJBHvhgXY0ggYGzD8GkHkHOqgJtSR9g-0g%3DW8tA%40mail.gmail.com#5e81023f1b5e918a41b8387ec18e82b3) (available in jdbc too). (think like redis pipeline)

This can be used by client drivers to groups queries in batches and send fewer RPC to the server backend.
